lib/zeitwerk/loader.rb

```diff
@@ -262,6 +266,23 @@ def collapse(*glob_patterns)
       end
     end
 
+    # Configure a block to be invoked once a certain constant path is loaded.
+    # Supports multiple callbacks, and if there are many, they are executed in
+    # the order in which they were defined.
+    #
+    #   loader.on_load("SomeApiClient") do
+    #     SomeApiClient.endpoint = "https://my-api.dev"
+    #   end
+    #
+    # @sig (String) { () -> void } -> void
+    def on_load(cpath, &block)
+      raise Error, "on_load only accepts strings" unless cpath.is_a?(String)
+
+      mutex.synchronize do
+        (on_load_callbacks[cpath] ||= []) << block
```

Comment:
You didn't seem so keen on it, but shouldn't we invoke the callback immediately if the module or class is already loaded? Or is there some adverse effects to that I'm not thinking of?

Reply:
It is a matter of how we want to define this feature. If you define an event on a button push, it does not matter if the button was pushed before.

In our case, furthermore, we do not know what we autoloaded if eager loading. A `const_defined?` is not scoped to the loader.

All in all, I think this logic is more clear, and enough for the use cases I have in mind, which happen before `setup`.

Reply:
> It is a matter of how we want to define this feature. If you define an event on a button push, it does not matter if the button was pushed before.

Agreed. But my definition of this feature is: "Execute this block as soon as this constant is loaded". Which would be consistent with `ActiveSupport.on_load`: https://api.rubyonrails.org/classes/ActiveSupport/LazyLoadHooks.html#method-i-on_load

 > In our case, furthermore, we do not know what we autoloaded if eager loading

I think we could just execute directly if eager loading.

Either way I can work around this limitation myself by using a higher level help, but I do really think that's what users expcetions and need would be.

Reply:
I guessed you had `AS::.on_load` in mind, but parity with that interface is not a goal of this feature per se. I need to think about what makes sense or is possible in this gem, and what does a callback mean in general. There are many callbacks that are only reactive, in JavaScript, in Active Record, in controllers, etc., so a reactive callback is a valid interface.

When I wrote "eager loading" I mean "reloading disabled", corrected it later, but too late for you to read it again.

In which use case you need to define a callback after `setup` and after the class has been loaded?

Reply:
> In which use case you need to define a callback after `setup` and after the class has been loaded?

Oh right I think we already had that discussion. I suppose that if we only allow to define these during setup, it's not as much as a big deal.

I'm just very worried about some callback accidentally not being fired anymore because of a subtle change in load order. But maybe I'm being too paranoid.

Reply:
Would it make sense to you to raise an error if someone tries to add a callback after the setup phase?

Reply:
I thought about that, but in this case I leant on being permissive. By not raising, we allow callbacks to be defined on reload according to file changes. Not that I have a use case in mind, just a gut feeling of not being rigid where I do not see an obvious reason to be so.

There's a bit of intuition in this decision and it is debatable, someone else (maybe you :), could perfectly decide to raise.

In the end, these trade-offs end up in the documentation/spec of the feature. The README tells you how it works, and it makes explicit that you can define after `setup`, it is your responsibility to know what you are doing.

Reply:
Ah, let me add that this flexibility has in mind that Zeitwerk is intended to be used by arbitrary Ruby projects, so Rails mental models should not constrain the design.

It is totally debatable, though.

Reply:
Fine by me. I guess we'll see how it goes anyway.

Reply:
Cool man, indeed it is tentative, time can perfectly prove it wrong.